app/recurring_expenses/views.py

```diff
@@ -39,53 +40,102 @@ def create_or_edit_recurring_expense(request, expense_id=None):
+                period=int(row[6]),
+                currency=row[4].strip(" "),
+            )
+            recurring_expense.save()
+            rows_imported += 1
+        except Exception as e:
+            rows_errored += 1
+            raise ValueError(str(e))
+
+
 @login_required
 def import_recurring_expenses(request):
     if request.method == "GET":
         return render(request, "recurring_expenses/import.html", {})
     if request.method == "POST":
-        uploaded_file = request.FILES.get("file")
+        try:
+            uploaded_file = request.FILES.get("file")
 
-        if uploaded_file is None:
-            messages.error(request, "No file uploaded")
-            return HttpResponseRedirect(reverse("recurring_expenses:import"))
+            if uploaded_file is None:
+                messages.error(request, "No file uploaded")
+                return HttpResponseRedirect(reverse("recurring_expenses:import"))
 
-        f = uploaded_file.read().decode("utf-8")
-        reader = csv.reader(f.splitlines(), delimiter=",")
-        next(reader, None)  # skip the headers
-
-        rows_imported = 0
-        rows_errored = 0
-        total_rows = 0
-
-        for row in reader:
-            if row[0].strip(" ").lower() == "totals":
-                break
-
-            total_rows += 1
-
-            try:
-                recurring_expense = RecurringExpense(
-                    active=parse_bool(row[2]),
-                    particulars=row[3].strip(" "),
-                    amount=row[4].strip("$  "),
-                    period=int(row[5]),
-                    currency=row[6].strip(" "),
-                )
-                recurring_expense.save()
-                rows_imported += 1
-            except Exception as e:
-                print(e)
-                rows_errored += 1
-
-        messages.success(
-            request,
-            "{}/{} Recurring expenses imported successfully. {} skipped".format(
-                rows_imported, total_rows, rows_errored
-            ),
-        )
+            if uploaded_file.content_type not in _IMPORT_CONTENT_TYPES:
+                messages.error(request, "Invalid file type")
+                return HttpResponseRedirect(reverse("recurring_expenses:import"))
 
-        return HttpResponseRedirect(reverse("recurring_expenses:import"))
+            if uploaded_file.content_type == _IMPORT_CONTENT_KEYS["json"]:
+                data = load_json_data(uploaded_file)
+                upload_json_date(data)
+            elif uploaded_file.content_type == _IMPORT_CONTENT_KEYS["csv"]:
+                reader = load_csv_data(uploaded_file)
+                upload_csv_data(reader)
+
+            # Redirect to a success page or do something else
+            messages.success(request, "Recurring expenses imported successfully")
+            return HttpResponseRedirect(reverse("recurring_expenses:import"))
+        except Exception as e:
+            messages.error(request, str(e))
```

Comment:
Ideally we should return something more user friendly to the user than the raw exception message. 

![image](https://github.com/iokiwi/moneyapp/assets/1480457/ed93ae8f-c4f8-4ab0-a1b1-0bc8f5f9ea04)

In transactions we handle integrity errors for Unique Key constraint violations by simply "skipping" the row and telling the user how many rows we skipped.

https://github.com/iokiwi/moneyapp/blob/main/app/transactions/views.py#L128

```
            except IntegrityError:
                skipped_rows += 1

        messages.success(
            request,
            "{}/{} transactions imported. {} duplicate transactions skipped.".format(
                rows_imported, len(ofx.account.statement.transactions), skipped_rows
            ),
        )
   ```

Reply:
Sure, I'll create a custom exceptions for the once we have raised and catch the ValueError for a more appropriate message in the next commit. :smile: